Async Programming.md

```diff
@@ -0,0 +1,15 @@
+1. - **Async Programming**: Asynchronous programming allows a program to perform tasks concurrently, improving efficiency and responsiveness. It often involves non-blocking operations that enable a program to continue executing other tasks while waiting for resources or data.
+    
+2. **Thread**: A thread is a lightweight, smaller unit of a process that can run concurrently with other threads within the same process. Threads share the same memory space and resources, making them efficient for parallel tasks but also prone to issues like race conditions.
```

Comment:
Explain race condition, how does it happened?

Reply:
occurs when multiple threads access shared resources concurrently and at least one thread modifies the resource. This can lead to unpredictable outcomes because the final state of the resource depends on the sequence or timing of the thread execution, which can vary each time the program runs.
To prevent race conditions, synchronization mechanisms such as mutexes, locks, or atomic operations are used to ensure that only one thread can access the critical section of code that modifies shared resources at a time.